File: agape-rs/src/ast.rs
//! The Agape abstract syntax tree (SPEC-1.0 §15.2).
//!
//! The shape the parser produces. It covers the whole v1.0 surface: the three
//! seams (cognition `<-`, tool calls, identity `verify … by p`), the graded
//! judgment type `Credence<E>`, the gates `decide`/`verify`, `quorum`/fusion
//! with `independent`/`dependent` declarations, user nominal types
//! (`struct`/`enum`/`event`), capabilities (`grants { emit/reach/use }`), the
//! reactive/query constructs, and the somatic kernel (`while`/`array`).

/// A type annotation. `event<T>` marks spine presence; `Credence<E>` is a graded
/// judgment over enum `E` (§3); `array<T>` is the somatic aggregate.
#[derive(Debug, Clone, PartialEq)]
pub enum Type {
    Int,
    Float,
    Bool,
    Text,
    Null,
    /// A user/prelude type referenced by name (enum, struct, agent, `Principal`,
    /// `Rule`, `Verification`, `Entailment`, ...). The name may be qualified
    /// (`m.Name`) or a generic type parameter (§19.2, §19.5).
    Named(String),
    /// A generic instantiation `Name<T, ...>` (a monomorphized struct, §19.5).
    Generic(String, Vec<Type>),
    /// `event<T>` — present (or to be present) on the spine.
    Event(Box<Type>),
    /// `array<T>` — the one primitive aggregate.
    Array(Box<Type>),
    /// `Credence<E>` — a graded distribution over enum `E`'s variants (§3).
    Credence(Box<Type>),
    /// `Decision<E>` — a gate's committed outcome over enum `E` (§3, settled).
    Decision(Box<Type>),
}

/// Binary operators. There is no similarity operator in v1.0 (§2).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
    Eq,
    Ne,
    Lt,
    Gt,
    Le,
    Ge,
}

impl BinOp {
    pub fn from_str(s: &str) -> Option<BinOp> {
        Some(match s {
            "+" => BinOp::Add,
            "-" => BinOp::Sub,
            "*" => BinOp::Mul,
            "/" => BinOp::Div,
            "==" => BinOp::Eq,
            "!=" => BinOp::Ne,
            "<" => BinOp::Lt,
            ">" => BinOp::Gt,
            "<=" => BinOp::Le,
            ">=" => BinOp::Ge,
            _ => return None,
        })
    }
    pub fn is_comparison(s: &str) -> bool {
        matches!(s, "==" | "!=" | "<" | ">" | "<=" | ">=")
    }
}

/// The parameter a gate decides by (`by` clause / `decide` rule, §3, §13).
/// `rule ::= cmpop Number ("margin" Number)? | expr`.
#[derive(Debug, Clone, PartialEq)]
pub enum GateBasis {
    /// `> 0.8` / `> 0.8 margin 0.1` / `confidence 0.8 margin 0.1` — a
    /// threshold/margin `Rule` written as sugar.
    Threshold { op: BinOp, value: f64, margin: f64 },
    /// `conformal α` — the conformal basis; calibrates from the spine, abstains
    /// below its readiness floor (§13).
    Conformal { alpha: f64 },
    /// A named `policy`, a `Rule` value, or (for `verify`/`attest … by p`) a `Principal`.
    Value(Box<Expr>),
}

/// Expressions.
#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Int(i64),
    Float(f64),
    Bool(bool),
    Null,
    Str(String),
    /// An f-string template (raw text incl. `{expr}` placeholders).
    FStr(String),
    Name(String),
    /// `self` — an agent's reference to itself.
    SelfRef,
    Binary { op: BinOp, left: Box<Expr>, right: Box<Expr> },
    /// Logical negation `!e`.
    Not(Box<Expr>),
    Call { func: Box<Expr>, args: Vec<Expr> },
    Member { obj: Box<Expr>, prop: String },
    Index { obj: Box<Expr>, index: Box<Expr> },
    /// `[a, b, c]` — array literal.
    Array(Vec<Expr>),
    /// `Name { field: v, ... }` — a struct literal (all fields required, §3).
    StructLit { name: String, fields: Vec<(String, Expr)> },
    /// `dest <- payload [expires N] [retry…]` — the one send operator (§6).
    Send {
        dest: Box<Expr>,
        payload: Box<Expr>,
        expires: Option<f64>,
        retry: Option<Box<RetryTail>>,
    },
    /// `source |> func` — concurrent fan-out over a collection (§12).
    Pipe { source: Box<Expr>, func: Box<Expr> },
    /// `decide(e, rule)` — the somatic gate-collapse `P → U` (§13). Rule mandatory.
    Decide { expr: Box<Expr>, rule: GateBasis },
    /// `e by rule` — collapse a `Credence<E>` to a `Decision<E>` (§13): settled,
    /// off-spine, in hand, *unendorsed* (not recorded). Rule mandatory.
    Collapse { expr: Box<Expr>, rule: GateBasis },
    /// `endorse(e by rule)` — the recorded gate in expression position (§13):
    /// collapse + record, yielding an *endorsed* `Decision<E>`.
    EndorseExpr { arg: Box<Expr>, rule: GateBasis },
    /// `verify gatearg [by basis]` — the recorded gate (§13): `decide` + emit.
    /// `by` absent ⇒ default `Rule`; `by` a `Principal` ⇒ identity-seam attest.
    Verify { arg: Box<Expr>, by: Option<GateBasis> },
    /// `quorum(k of c1, …, cn)` — graded "at least k of n commit" (§12).
    Quorum { k: i64, judges: Vec<Expr> },
    /// A query used in expression position (yields its result set; lands nothing).
    Query(Box<Query>),
}

/// The three memory/spine query forms (§10).
#[derive(Debug, Clone, PartialEq)]
pub enum Query {
    /// `find BINDING [, origin(BINDING)] where { S P O; ... }` — graph query.
    Find { binding: String, origin: Option<String>, pattern: Vec<(String, String, String)> },
    /// `select COLS from SOURCE where { col op value, ... }` — fact/spine scan.
    Select { cols: Vec<String>, source: String, star: bool, conds: Vec<Cond> },
    /// `match { BINDING: QUERY } > THRESHOLD` — vector store; a gate (§10).
    Match { binding: String, query: Box<Expr>, threshold: f64 },
}

/// A `where`-clause condition: `col op value` or `col : value` (op == ":").
#[derive(Debug, Clone, PartialEq)]
pub struct Cond {
    pub col: String,
    pub op: String,
    pub value: Expr,
}

/// A grant entry kind (§13): emit an event, reach an agent, or use a tool.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CapKind {
    Emit,
    Perform,
    Reach,
    Use,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Capability {
    pub kind: CapKind,
    pub target: String,
}

/// An `interface` member (§19.5): a handled-event→decision contract, or a
/// required capability.
#[derive(Debug, Clone, PartialEq)]
pub enum IfaceMember {
    /// `when EVENT decide RESULT` — the agent must have a `when (EVENT …)` handler.
    When { event: String, result: String },
    /// `requires CAP` — the capability must be in the implementor's `grants`.
    Requires(Capability),
}

/// A formal parameter (always `type name`).
#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub ty: Type,
    pub name: String,
}

/// A struct/event field (`name: T` or `T name`).
#[derive(Debug, Clone, PartialEq)]
pub struct Field {
    pub name: String,
    pub ty: Type,
}

/// Fusion dependence declaration (§12).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Dep {
    Independent,
    Dependent,
}

/// The tail of a send-retry, or a standalone `retry` block (§11).
#[derive(Debug, Clone, PartialEq)]
pub enum RetryTail {
    /// `retry(N) { body }` — bounded; a handler runs before each re-attempt.
    Bounded { count: i64, body: Vec<Stmt> },
    /// `retry(TYPE x: PRED) { body }` — unbounded predicate form (Turing complete).
    Predicate { ty: Type, bind: String, pred: Expr, body: Vec<Stmt> },
}

/// A tool's effect class (§6b): `read` observes the world (its result carries its
/// inputs' trust); `write` changes the world (a consequential sink whose inputs must
/// be settled). The class is mandatory on every `tool` declaration — never defaulted.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ToolEffect {
    Read,
    Write,
}

/// Statements (and declarations — both are `decl`/`stmt` in §15.2).
#[derive(Debug, Clone, PartialEq)]
pub enum Stmt {
    /// `TYPE NAME [= EXPR];`
    VarDecl { ty: Type, name: String, expr: Option<Expr> },
    /// `[sync] RET NAME[<typarams>](params) { body }`.
    FnDecl { is_sync: bool, ret: Type, name: String, typarams: Vec<String>, params: Vec<Param>, body: Vec<Stmt> },
    /// `[pub] agent NAME [(params)] [: Iface, ...] [grants {...}] { body }`.
    AgentDecl {
        name: String,
        params: Vec<Param>,
        /// Implemented interfaces, nominal (`agent A : I, J`, §19.5).
        ifaces: Vec<String>,
        grants: Vec<Capability>,
        has_grants: bool,
        /// `grants { * }` — the explicit unconstrained opt-out (lattice top).
        unconstrained: bool,
        body: Vec<Stmt>,
    },
    /// `[pub] struct NAME[<typarams>] { field, ... }`.
    StructDecl { name: String, typarams: Vec<String>, fields: Vec<Field> },
    /// `enum NAME { A, B, ... }` — a closed variant set (no payloads in v1.0).
    EnumDecl { name: String, variants: Vec<String> },
    /// `event NAME(field, ...) [: Super];` — a custom spine-event type with typed
    /// payload; `error_super` ⇒ a leaf under the built-in `Error` root (§19.5).
    /// `super_name` carries any declared supertype spelling (for the non-`Error`
    /// rejection); `None` if no supertype.
    EventDecl { name: String, fields: Vec<Field>, error_super: bool, super_name: Option<String> },
    /// `interface NAME { (when EVENT decide RESULT | requires CAP);* }` (§19.5).
    InterfaceDecl { name: String, members: Vec<IfaceMember> },
    /// `[reversible] action NAME(field, ...);` — a performative event type (§3, §13);
    /// `perform`ing it is a consequential act needing the `perform NAME` power and a
    /// settled value. `reversible` marks a low-stakes sink (§20).
    ActionDecl { name: String, fields: Vec<Field>, reversible: bool },
    /// `[reversible] read|write tool RET NAME(params);` — a tool-seam capability (§6b).
    /// The effect class is mandatory; `reversible` marks a low-stakes write sink (§20).
    ToolDecl { name: String, params: Vec<Param>, ret: Option<Type>, effect: ToolEffect, reversible: bool },
    /// `conformal α;` — a file-level default conformal guarantee (§20).
    ConformalDecl(f64),
    /// `authority EventType;` — marks an event type consequential (§13).
    Authority(String),
    /// `principal NAME;` — an identity-seam principal binding (§13).
    Principal(String),
    /// `extend PARENT(args);` — composition/inheritance (first stmt of an agent).
    Extend { parent: String, args: Vec<Expr> },
    /// `spawn TYPE name (args)?;` — allocate + construct (args bound here, §5).
    Spawn { agent_type: String, name: String, args: Vec<Expr> },
    /// `awake NAME [(args)];`
    Awake { name: String, args: Vec<Expr> },
    /// `sleep NAME;`
    Sleep(String),
    /// `prompt TYPE name;` — an external input sensor (§5b).
    Prompt { ty: Type, name: String },
    /// `verify gatearg [by basis];` — the recorded gate (§13).
    Verify { arg: Expr, by: Option<GateBasis> },
    /// `emit EventType(payload);`
    Emit { event_type: String, payload: Expr },
    /// `endorse (arg by rule) { arms } [abstain { ... }]` — the recorded gate (§13):
    /// collapse the `Credence`, append `Decided`/`Abstained`, dispatch the arms.
    Endorse { arg: Expr, rule: GateBasis, arms: Vec<(String, Vec<Stmt>)>, abstain: Option<Vec<Stmt>> },
    /// `perform ActionType(payload);` — a consequential act (§13).
    Perform { action_type: String, payload: Expr },
    /// `attest e by PRINCIPAL (arms | ;)` — the recorded identity-seam gate (§13):
    /// reach the principal, record an `Attestation`/`FailedAttestation`. `by` is an
    /// expression so `by "alice"` parses (the checker rejects text→Principal, §3).
    Attest { arg: Expr, by: Expr, arms: Vec<(String, Vec<Stmt>)> },
    /// `policy NAME { threshold θ  margin δ  floor m  conformal α  readiness N  fallback p }`
    /// — a named decision-policy bundle (§13), source-level, not config.
    PolicyDecl {
        name: String,
        threshold: Option<f64>,
        margin: Option<f64>,
        floor: Option<f64>,
        conformal: Option<f64>,
        readiness: Option<i64>,
        fallback: Option<String>,
    },
    /// `{ ... }` — a bare lexical block (the body of a `{ block } retry(N)`, or a
    /// nested scope). Carries no looping by itself; `retry` wraps it.
    Block(Vec<Stmt>),
    /// `say(EXPR);`
    Say(Expr),
    /// `return [EXPR];`
    Return(Option<Expr>),
    /// `if (cond) { then } [else { else }]`
    If { cond: Expr, then_body: Vec<Stmt>, else_body: Vec<Stmt> },
    /// `TARGET = EXPR;`
    Assign { target: Expr, expr: Expr },
    /// `on awake { ... }` / `on sleep { ... }` / `on crash { ... }`
    On { event: String, body: Vec<Stmt> },
    /// `when (Type binder? [about subj]) [if (guard)] { ... }` — a prospective,
    /// typed subscription (§7). `binder` evaluates to the matched event's payload.
    When { ty: Type, binder: Option<String>, about: Option<Expr>, guard: Option<Expr>, body: Vec<Stmt> },
    /// `catch [EventType][(SUBJECT)] as BINDING { ... }`
    Catch { event_type: Option<String>, subject: Option<Expr>, binding: String, body: Vec<Stmt> },
    /// `case (EXPR) as BINDING { Variant: {..} ... [default: {..}] }`.
    Case { expr: Expr, binding: String, arms: Vec<(String, Vec<Stmt>)>, default: Option<Vec<Stmt>> },
    /// `retry(N) { ... }` / `retry(TYPE x: PRED) { ... }` — standalone block form.
    Retry(RetryTail),
    /// `independent a, b, c;` / `dependent a, b;` — fusion dependence (§12).
    DepDecl { kind: Dep, names: Vec<String> },
    /// A query in statement position (lands `QueryResult(subject)`, §10).
    QueryStmt(Query),
    /// `while (COND) { BODY }` — the imperative loop (somatic kernel).
    While { cond: Expr, body: Vec<Stmt> },
    /// `break;`
    Break,
    /// `module a.b;` — the module-path header (§19.2).
    ModuleDecl { path: String },
    /// `[pub] import m [as x];` / `[pub] import { A, B } from m;` (§19.2, §19.2a).
    Import {
        /// The module path imported from.
        module: String,
        /// `Some(alias)` for `import m as x;` (whole-module prefix rebind); `None`
        /// for a plain whole-module import or a selective import.
        alias: Option<String>,
        /// Selective names for `import { A, B } from m;`; empty ⇒ whole-module.
        names: Vec<String>,
        /// `pub import …` re-exports (§19.2a).
        is_pub: bool,
    },
    /// `[principal] decide EXPR [conformal α] { Variant: … default: … } [defer to p];`
    /// — the readable gate (§20). Desugars to the §13 engine.
    Decide {
        /// The collapsed `Credence` expression.
        expr: Expr,
        /// A principal subject (`p decide …`), if present.
        subject: Option<String>,
        /// A per-gate conformal α override.
        conformal: Option<f64>,
        /// The commit arms.
        arms: Vec<(String, Vec<Stmt>)>,
        /// The optional `default:` arm.
        default: Option<Vec<Stmt>>,
        /// A trailing `defer to p` principal, if present.
        defer_to: Option<String>,
    },
    /// `#!somatic` / `#!cognitive` — a module attribute.
    ModuleAttr { name: String },
    /// A `pub`-prefixed declaration (§19.4) — exports the wrapped decl's name. The
    /// inner statement is the actual declaration; `pub` only affects cross-module
    /// name visibility, never runtime behavior.
    Pub(Box<Stmt>),
    /// A bare expression statement (e.g. a send, a tool call, a query expr).
    ExprStmt(Expr),
}

impl Stmt {
    /// The declared name of a top-level declaration, if it has one (for module
    /// resolution / visibility, §19). Unwraps a `pub` wrapper.
    pub fn decl_name(&self) -> Option<&str> {
        match self {
            Stmt::Pub(inner) => inner.decl_name(),
            Stmt::FnDecl { name, .. }
            | Stmt::AgentDecl { name, .. }
            | Stmt::StructDecl { name, .. }
            | Stmt::EnumDecl { name, .. }
            | Stmt::EventDecl { name, .. }
            | Stmt::ActionDecl { name, .. }
            | Stmt::ToolDecl { name, .. }
            | Stmt::InterfaceDecl { name, .. } => Some(name),
            _ => None,
        }
    }
}

File: agape-ts/src/ast.ts
// AST — the abstract syntax produced by the parser (a v0 subset of SPEC.md §15.2).

import type { Pos } from "./lexer.js";

export interface Node {
  pos: Pos;
}

// ---- Types ----
export type TypeRef =
  | { kind: "scalar"; name: "int" | "float" | "bool" | "text" | "null" }
  | { kind: "mem" } // a private-memory handle (§10)
  | { kind: "event"; inner: TypeRef } // legacy/deprecated: surface `event<T>` is rejected by the parser
  | { kind: "array"; inner: TypeRef }
  | { kind: "credence"; enumName: string }
  | { kind: "decision"; enumName: string }
  | { kind: "endorsement"; inner: TypeRef }
  | { kind: "named"; name: string; typeArgs?: TypeRef[] }; // enum/struct/agent/action/Principal/LedgerEntry; typeArgs at a generic-instantiation site (§19.5)

// ---- Program & declarations ----
export interface Program extends Node {
  kind: "program";
  // an optional leading `module modpath;` header (§19.2). A dotted name (e.g. "explicit.root");
  // undefined = the implicit root module (a file with no header and no import). Backward-compatible:
  // an unmodified single-module program leaves it undefined and is byte-identical to before.
  module?: string;
  // the header `import*` block (§19.2/§19.2a). Absent/empty for a program with no imports.
  imports?: ImportDecl[];
  decls: Decl[];
  stmts: Stmt[]; // top-level statements (spawn/awake/...)
}

// An `import` header (§19.2/§19.2a). `pub import` re-exports (whole prefix or selective). `selective`
// present (a name list) = `import { A, B } from M`; absent = a whole-module import `import M (as X)?`.
export interface ImportDecl extends Node {
  kind: "import";
  pub: boolean; // `pub import …` re-exports into this module's public surface (§19.2a)
  module: string; // the imported module's (possibly dotted) modpath
  selective?: string[]; // the selectively-imported bare names, when `import { … } from M`
  alias?: string; // the rebound prefix, when `import M as X` (whole-module only)
}

export type Decl = EnumDecl | StructDecl | ActionDecl | EventDecl | AgentDecl | InstructionDecl | ToolDecl | FnDecl | InterfaceDecl | PrincipalDecl | PromptDecl | ConfDecl | PolicyDecl;

// `principal NAME config?;` (§3, grammar §15.2 line 1463) — declares an accountable identity, a
// declared dependency bound to an identity backend by config (§17.1). Opaque, unforgeable,
// declared-not-constructed; used at a `decide c by NAME` (or the prefix `NAME decide …`) site (§13).
export interface PrincipalDecl extends Node {
  kind: "principal";
  name: string;
}
// `prompt TYPE NAME;` (§5b, grammar §15.2 line 1459) — opens a standing external input sensor. Its
// arrivals land `Prompt` events subjected `NAME`; `when (Prompt p about NAME)` subscribes. TYPE-first.
export interface PromptDecl extends Node {
  kind: "prompt";
  type: TypeRef;
  name: string;
}
// `conformal Number;` (grammar §15.2 line 1416 confdecl) — the file-level default conformal α (§20).
// Parse-only in v0: it carries the default α but drives no runtime effect (per-gate `by conformal α`
// still overrides it at the gate).
export interface ConfDecl extends Node {
  kind: "conformal";
  alpha: number;
}
// `policy NAME { threshold θ | margin δ | floor m }` (§13) — a named, reusable decision rule bundle in
// SOURCE (never the manifest, §17.2). `decide c by NAME` applies it: commit iff score ≥ threshold AND margin
// ≥ δ AND margin ≥ floor; a margin below `floor` abstains (the typed trigger for escalation, §13).
export interface PolicyDecl extends Node {
  kind: "policydecl";
  name: string;
  threshold?: number;
  margin?: number;
  floor?: number;
}

export interface EnumDecl extends Node {
  kind: "enum";
  name: string;
  variants: string[];
  pub?: boolean;
}
export interface StructDecl extends Node {
  kind: "struct";
  name: string;
  fields: Field[];
  typarams?: string[]; // type parameters `<A, B, …>` — empty/absent when non-generic (§19.5)
  pub?: boolean;
}
export interface Field {
  type: TypeRef;
  name: string;
}
export interface ActionDecl extends Node {
  kind: "action";
  name: string;
  fields: Field[];
  reversible: boolean;
  pub?: boolean;
}
export interface EventDecl extends Node {
  kind: "event";
  name: string;
  fields: Field[];
  errorSuper: boolean; // `: Error` — a valid Error-supertype leaf (§19.5)
  superName?: string;  // the raw supertype ident as written, when a supertype was declared (`: X`).
                       // A non-`Error` supertype is a TypeError diagnosed in the checker (§19.5/§19.6),
                       // NOT a ParseError — the parser only records the name.
  pub?: boolean;
}
export interface InstructionDecl extends Node {
  kind: "instruction";
  text: string;
}
// A tool — the world dependency (§6b). The effect class is mandatory.
export interface ToolDecl extends Node {
  kind: "tool";
  effect: "read" | "write";
  reversible: boolean;
  ret: TypeRef;
  name: string;
  params: Field[];
  pub?: boolean;
}
// A minimal function declaration (§15.2). Only `sync`-color checking is modeled here.
export interface FnDecl extends Node {
  kind: "fn";
  sync: boolean;
  ret: TypeRef;
  name: string;
  params: Field[];
  body: Stmt[];
  typarams?: string[]; // type parameters `<A, …>` — empty/absent when non-generic (§19.5)
  pub?: boolean;
}
// An interface names an agent's external surface (§19.5): the events it handles and the outcome each
// produces, plus the powers it `requires`. A TYPE (usable as a binding/param/reach target) but NOT
// instantiable. Not generic. Erases after checking.
export interface InterfaceDecl extends Node {
  kind: "interface";
  name: string;
  members: IfMember[];
  pub?: boolean;
}
export type IfMember =
  | { kind: "handler"; event: TypeRef; outcome: TypeRef } // `when EVENT decide RESULT`
  | { kind: "requires"; cap: "perform" | "reach" | "use"; name: string }; // `requires cap NAME`
export interface AgentDecl extends Node {
  kind: "agent";
  name: string;
  grants: Grant[] | "all"; // [] = default-deny; "all" = grants { * }
  params: Field[]; // constructor params `agent Boss(Worker hand)` — bound at spawn (E-Spawn §15.4.2)
  fields: Field[];
  hooks: OnHook[];
  whens: WhenStmt[];
  instructions: string[];
  ctor: Stmt[]; // the agent-body constructor statements (run at spawn, E-Spawn §15.4.2)
  extends?: { name: string; args: Expr[] }; // `extend NAME(args);` — subtractive inheritance (§5, §13)
  ifaces?: string[]; // implemented interfaces (`: Iface, …`) — nominal conformance (§19.5)
  pub?: boolean;
}
export type Grant =
  | { cap: "perform"; name: string }
  | { cap: "reach"; name: string }
  | { cap: "use"; name: string };

// ---- Statements ----
export type Stmt =
  | VarDecl
  | AssignStmt
  | SpawnStmt
  | AwakeStmt
  | SleepStmt
  | EmitStmt
  | PerformStmt
  | SayStmt
  | ReturnStmt
  | IfStmt
  | DispatchStmt
  | WhenStmt
  | MemDecl
  | ForgetStmt
  | DepDeclStmt
  | RetryStmt
  | ExprStmt;

// `{ block } retry(N)` (§11, §15.2) — the ONLY loop, bounded: re-attempt the block up to N times on a fault
// (an `Error`, e.g. a `TypeMismatch` from a malformed reply); on exhaustion, emit `RetryExhausted` and the
// fault propagates. Every Agape program terminates because this is the sole loop and N is a literal bound.
export interface RetryStmt extends Node {
  kind: "retry";
  body: Stmt[];
  n: number;
}

// `("independent"|"dependent") Ident ("," Ident)* ";"` — declare the dependence structure of the
// values fused by `all`/`any`/`quorum` (§12, §15.2 depdecl). A static assertion consumed by the
// T-Fuse coverage check; it produces no runtime effect and no ledger event of its own.
export interface DepDeclStmt extends Node {
  kind: "depdecl";
  relation: "independent" | "dependent";
  names: string[];
}

// `mem NAME [<- EXPR];` — declare a private-memory handle, optionally initialized (§10).
export interface MemDecl extends Node {
  kind: "memdecl";
  name: string;
  init?: Expr;
}
// `forget NAME;` — tombstone a `mem` handle; consumes it (§10).
export interface ForgetStmt extends Node {
  kind: "forget";
  name: string;
}

export interface VarDecl extends Node {
  kind: "var";
  type: TypeRef;
  name: string;
  init?: Expr;
}
export interface AssignStmt extends Node {
  kind: "assign";
  target: Expr; // postfix lvalue
  value: Expr;
}
export interface SpawnStmt extends Node {
  kind: "spawn";
  agentType: string;
  name: string;
  args: Expr[];
}
export interface AwakeStmt extends Node {
  kind: "awake";
  name: string;
}
export interface SleepStmt extends Node {
  kind: "sleep";
  name: string;
}
export interface EmitStmt extends Node {
  kind: "emit";
  name: string;
  args: Expr[];
}
export interface PerformStmt extends Node {
  kind: "perform";
  name: string;
  args: Expr[];
}
export interface SayStmt extends Node {
  kind: "say";
  arg: Expr;
}
export interface ReturnStmt extends Node {
  kind: "return";
  value?: Expr;
}
export interface IfStmt extends Node {
  kind: "if";
  cond: Expr;
  then: Stmt[];
  else?: Stmt[];
}
export interface OnHook extends Node {
  kind: "on";
  event: "awake" | "sleep" | "crash";
  body: Stmt[];
}
export interface WhenStmt extends Node {
  kind: "when";
  etype: string;
  binder?: string;
  about?: Expr;
  guard?: Expr;
  body: Stmt[];
}

// gate sugar: `gate armblock (abstain block)?`
export interface DispatchStmt extends Node {
  kind: "dispatch";
  gate: GateExpr;
  arms: Arm[];
  abstain?: { binder?: string; body: Stmt[] };
}
export interface Arm {
  head: string; // a variant name, or "true"/"false"
  binder?: string; // `as e`
  body: Stmt[];
}
export interface ExprStmt extends Node {
  kind: "exprstmt";
  expr: Expr;
}

// ---- Expressions ----
export type Expr =
  | GateExpr
  | SendExpr
  | RecallExpr
  | SelectExpr
  | FindExpr
  | MatchExpr
  | StructLit
  | FStringExpr
  | StringLit
  | IntLit
  | FloatLit
  | BoolLit
  | NullLit
  | SelfExpr
  | IdentExpr
  | CallExpr
  | MemberExpr
  | IndexExpr
  | BinaryExpr
  | UnaryExpr
  | AggExpr
  | QuorumExpr
  | PipeExpr
  | ArrayLit;

// `("all"|"any") "(" expr ("," expr)* ")"` — reduce a comma-list of operands OR a single
// `array<Credence<bool>>` (§12, §15.2). Over plain `bool` it is ordinary conjunction/disjunction;
// over `Credence<bool>` it fuses evidence into a single `Credence<bool>` under the declared structure.
export interface AggExpr extends Node {
  kind: "agg";
  op: "all" | "any";
  operands: Expr[];
}
// `"quorum" "(" Int "," expr ")"` — "at least k of n commit," a thresholded reduction over the same
// fusion algebra as all/any (§12, §15.2).
export interface QuorumExpr extends Node {
  kind: "quorum";
  k: number;
  source: Expr;
}
// `expr "|>" expr` — pipe each element of a collection through `fn` (§12, §15.2).
export interface PipeExpr extends Node {
  kind: "pipe";
  source: Expr;
  fn: Expr;
}
// `"[" (expr ("," expr)*)? "]"` — an array literal (§15.2 primary), consumed by |>/all/any/quorum.
export interface ArrayLit extends Node {
  kind: "arraylit";
  items: Expr[];
}

// `MEM -> "query"` — recall from a private-memory handle (§10); always tainted.
export interface RecallExpr extends Node {
  kind: "recall";
  mem: Expr;
  query: Expr;
}
// `select COLS from TARGET where { COND }`, or `select Event as e from ledger where { e.field ... }`
// — a facts/ledger query (§10).
export interface SelectExpr extends Node {
  kind: "select";
  cols: string[] | "*";
  target: string; // an agent instance name, or `ledger`/`self`
  eventType?: string; // new ledger-row form: `select Held as h from ledger ...`
  alias?: string;
  cond: QueryCond[];
}
// `find x [, origin(x)] where { TRIPLE+ }` — a relationship-graph query (§10).
export interface FindExpr extends Node {
  kind: "find";
  binder: string;
  origin: boolean;
  triples: Triple[];
}
// `match VECTOR > θ` — a vector-store similarity query; a gate (§10).
export interface MatchExpr extends Node {
  kind: "match";
  vector: Expr;
  theta: number;
}
export interface StructLit extends Node {
  kind: "structlit";
  typeName?: string;
  typeArgs?: TypeRef[]; // `Box<int> { value: 7 }` — parsed for the non-generic check, erased past it (§19.5)
  fields: { name: string; value: Expr }[];
}
export interface QueryCond {
  field: string;
  op: string; // ==, !=, <, >, <=, >=  (a `:` desugars to ==)
  value: Expr;
  connective?: "&&" | "||"; // the connective JOINING this cond to the previous one
}
export interface Triple {
  subject: string;
  predicate: string;
  object: string;
}

export type GateExpr = DecideExpr | EndorseExpr;

export type Rule =
  // inline rules carry their optional consequential `floor m` (the runtime margin floor, §13) and,
  // for conformal, an optional `readiness N` (the labelled-case minimum before autonomous commit).
  | { kind: "confidence"; theta: number; margin?: number; floor?: number }
  | { kind: "conformal"; alpha?: number; readiness?: number; floor?: number }
  | { kind: "policy"; name: string }
  | { kind: "expr"; expr: Expr };

export interface DecideExpr extends Node {
  kind: "decide";
  principal?: string; // optional escalation prefix (a declared `principal` NAME)
  principalStr?: boolean; // the prefix was a STRING literal, not an ident — a TypeError (§3, §13): a principal
  //                         basis must be a declared `principal`, not a forgeable text claim.
  credence: Expr;
  rule: Rule;
}
export interface EndorseExpr extends Node {
  kind: "endorse";
  subject: Expr;
  decision: Expr;
}
export interface SendExpr extends Node {
  kind: "send";
  dest: Expr;
  message: Expr;
  expires?: number;
}
export interface FStringExpr extends Node {
  kind: "fstring";
  parts: ({ kind: "text"; text: string } | { kind: "expr"; expr: Expr })[];
}
export interface StringLit extends Node { kind: "string"; value: string; }
export interface IntLit extends Node { kind: "int"; value: number; }
export interface FloatLit extends Node { kind: "float"; value: number; }
export interface BoolLit extends Node { kind: "bool"; value: boolean; }
export interface NullLit extends Node { kind: "null"; }
export interface SelfExpr extends Node { kind: "self"; }
export interface IdentExpr extends Node { kind: "ident"; name: string; }
export interface CallExpr extends Node { kind: "call"; callee: Expr; args: Expr[]; }
export interface MemberExpr extends Node { kind: "member"; obj: Expr; field: string; }
export interface IndexExpr extends Node { kind: "index"; obj: Expr; index: Expr; } // `a[i]` element access (§10)
export interface BinaryExpr extends Node { kind: "binary"; op: string; left: Expr; right: Expr; }
export interface UnaryExpr extends Node { kind: "unary"; op: string; operand: Expr; }
